plugins/dbus_announce.c

```diff
@@ -0,0 +1,143 @@
+}
+
+static rpmRC open_dbus(rpmPlugin plugin, rpmts ts)
+{
+    struct stat st;
+    DBusError err;
+    int rc = 0;
+    struct dbus_announce_data * state = rpmPluginGetData(plugin);
+
+    /* Assume we are logging but... */
+    state->logging = 1;
+
+    /* ...don't log test transactions */
+    if (rpmtsFlags(ts) & (RPMTRANS_FLAG_TEST|RPMTRANS_FLAG_BUILD_PROBS))
+	state->logging = 0;
+
+    /* ...don't log chroot transactions */
+    if (!rstreq(rpmtsRootDir(ts), "/"))
+	state->logging = 0;
+
+    /* Don't open */
+    if (!state->logging || state->bus)
+	return RPMRC_OK;
+
+    if (lstat("/run/systemd/system/", &st) == 0) {
+        if (S_ISDIR(st.st_mode)) {
+	    // initialise the error value
+	    dbus_error_init(&err);
+
+	    state->bus = dbus_bus_get_private(DBUS_BUS_SYSTEM, &err);
+	    if (dbus_error_is_set(&err)) {
+		fprintf(stderr, "Connection Error (%s)\n", err.message);
+		dbus_error_free(&err);
+	    }
+	    if (state->bus) {
+		rc = dbus_bus_request_name(state->bus, "org.rpm.announce",
+				  DBUS_NAME_FLAG_REPLACE_EXISTING , &err);
+	    }
+	    if (dbus_error_is_set(&err)) {
+		//fprintf(stderr, "Name Error (%s)\n", err.message);
+		dbus_error_free(&err);
+	    }
+	    if (DBUS_REQUEST_NAME_REPLY_PRIMARY_OWNER != rc) {
+		dbus_connection_close(state->bus);
+		dbus_connection_unref(state->bus);
+		state->bus = NULL;
+		state->logging = 0;
+		return RPMRC_NOTFOUND;
+	    }
+            return RPMRC_OK;
+        }
+    }
+
+    return RPMRC_NOTFOUND;
+}
+
+static void dbus_announce_cleanup(rpmPlugin plugin)
+{
+    struct dbus_announce_data * state = rpmPluginGetData(plugin);
+    if (state->bus) {
+	dbus_connection_close(state->bus);
+	dbus_connection_unref(state->bus);
+    }
+    free(state);
+}
+
+static rpmRC send_ts_message(rpmPlugin plugin,
+			     const char * name,
+			     rpmts ts,
+			     int res)
+{
+    struct dbus_announce_data * state = rpmPluginGetData(plugin);
+    DBusMessage* msg;
+    dbus_uint32_t serial = 0;
+    char * dbcookie;
+
+    if (!state->logging) {
+	return RPMRC_OK;
+    }
+
+    msg = dbus_message_new_signal("/org/rpm/Transaction", // object name
+				  "org.rpm.Transaction", // interface name
+				  name); // name of the signal
+    if (NULL == msg) {
+	return RPMRC_FAIL;
+    }
+
+    dbcookie = rpmdbCookie(rpmtsGetRdb(ts));
+    if (!dbus_message_append_args(msg, DBUS_TYPE_STRING, &dbcookie)) {
+	return RPMRC_FAIL;
+    }
+    dbcookie = _free(dbcookie);
+
+    if (!dbus_connection_send(state->bus, msg, &serial)) {
+	return RPMRC_FAIL;
+    }
+    dbus_connection_flush(state->bus);
+    return RPMRC_OK;
+}
+
+static rpmRC dbus_announce_tsm_pre(rpmPlugin plugin, rpmts ts)
+{
+    int rc;
+
+    rc = open_dbus(plugin, ts);
+    if (rc != RPMRC_OK)
+	return rc;
+    return send_ts_message(plugin, "StartTransaction", ts, RPMRC_OK);
+}
+
+static rpmRC dbus_announce_tsm_post(rpmPlugin plugin, rpmts ts, int res)
+{
+    return send_ts_message(plugin, "CompleteTransaction", ts, res);
```

Comment:
This is perhaps in the bike-shedding department, but "complete" by definition seems to indicate success, whereas this can have an error condition as well. So maybe "EndTransaction" would be closer to pairing with Start in these context.

Reply:
Using EndTransaction instead.